content/technotes/2018-09-09-github-badges.md

```diff
@@ -0,0 +1,395 @@
+  github_repo <- stringr::str_remove(github_repo, "\\#.*$")
+  github_repo <- stringr::str_remove(github_repo, "\\#.*[ \\(.*\\)]")
+  github_repo <- stringr::str_remove(github_repo, "/$")
+  stringr::str_replace(github_repo,".*\\.com\\/", "")
+}
+
+
+github_cran <- dplyr::group_by(github_cran, Package)
+github_cran <- dplyr::mutate(github_cran, github = select_github_repo(URL))
+github_cran <- tidyr::separate(github_cran, github, "\\/", 
+                               into = c("owner", "repo"))
+github_cran <- dplyr::ungroup(github_cran)
+
+# Not very general
+github_cran$repo[which(github_cran$Package == "webp")] <- "webp"
+```
+
+I needed a bit of string cleaning mostly to deal with the URLs of
+[Jeroen Ooms](https://ropensci.org/about/#team)’ packages, see e.g.
+[this one](https://cran.r-project.org/web/packages/antiword/index.html).
+I guess I could have cleaned even more, but it was good enough for this
+exploration.
+
+### Get all badges
+
+Then for each repo I queried the download URL of the [preferred
+README](https://developer.github.com/v3/repos/contents/#get-the-readme)
+via GitHub’s V3 API, using [the `gh`
+package](https://github.com/r-lib/gh). The preferred README is the one
+GitHub displays on the repo landing page. I used
+`codemetar::extract_badges`, of course. I rate-limited the basic
+function using [`ratelimitr`](https://github.com/tarakc02/ratelimitr).
+
+``` r
+library("magrittr")
+
+github_cran <- readr::read_csv("data/github_cran_links.csv")
+
+.get_badges <- function(owner, repo){
+  message(paste(owner, repo, sep = "/"))
+  readme <- try(gh::gh("GET /repos/:owner/:repo/readme",
+                       owner = owner, repo = repo),
+                silent = TRUE)
+  if(inherits(readme, "try-error")){
+    return(NULL)
+  }else{
+    
+    badges <- codemetar::extract_badges(readme$download_url)
+    
+    if(nrow(badges)>0){
+      badges$owner <- owner
+      badges$repo <- repo
+    }
+    
+    return(badges)
+  }
+  
+  
+}
+
+.get_badges %>%
+  ratelimitr::limit_rate(ratelimitr::rate(1, 1)) -> get_badges
+
+purrr::map2_df(github_cran$owner, 
+               github_cran$repo,
+              get_badges) %>%
+  readr::write_csv(path = "data/aaall_badges.csv")
+```
+
+Don’t judge me by my filenaming skills. I was *maybe* a bit too
+enthusiastic!
+
+Analyze badges from CRAN packages
+---------------------------------
+
+I wanted to answer several questions about the badges of CRAN packages,
+beyond being just happy to have been able to collect so many of them.
+
+### How many repos have at least one badge?
+
+``` r
+github_cran <- readr::read_csv("data/github_cran_links.csv")
+# the same repo can have been used by several packages!
+badges <- readr::read_csv("data/aaall_badges.csv")
+badges <- dplyr::distinct(badges)
+
+nobadges <- dplyr::anti_join(github_cran, badges,
+                             by = c("owner", "repo"))
+```
+
+There are 1237 packages without any badge (or rather said, without any
```

Comment:
nice! Can we give these as percents?  Might be worth noting the total sample size that has GitHub links as a fraction of CRAN packages, just for fun.

Reply:
Both excellent points!